File: Chapter 6 - Lecture/p11.m
%P11  Chebyshev differentiation of a smooth function
close all;
scrsz = get(groot,'ScreenSize');
figure('position', [150   100   0.8*scrsz(3:4)]); 

%%
xx = -1:0.01:1;
uu = exp(xx).*sin(5*xx); clf
for N = [10 20]
    % MOST INTERESTING CODE
    %==================================
    [D, x] = cheb(N);
    u = exp(x).*sin(5*x);
    %==================================
    subplot(2,2, 2*(N == 20)+1)
    plot(x, u, '.', 'markersize', 14), grid on
    line(xx, uu, 'linewidth', 0.8, 'color','k');
    title(['u(x), N=', int2str(N)]);
    % MOST INTERESTING CODE
    %==================================
    error = D*u - exp(x).*(sin(5*x)+5*cos(5*x));
    %==================================
    subplot(2,2, 2*(N == 20)+2)
    plot(x, error, '.', 'markersize', 14), grid on
    line(x, error, 'linewidth', 0.8, 'color','k')
    title([' error in u''(x), N=' int2str(N)])
end

File: p11.py
"""
Chebyshev differentiation of a smooth function

@author: Khiem Nguyen
"""

import numpy as np
import matplotlib.pyplot as plt
from cheb import cheb

xx = np.arange(-1.0, 1.01, 0.01)
uu = np.exp(xx) * np.sin(5*xx)
for N in [10, 20]:
    D, x = cheb(N)
    u = np.exp(x) * np.sin(5*x)
    plt.figure(figsize=(10,8))
    plt.subplot(position=[0.15, 0.66 - 0.4*(N==20), 0.31, 0.28])
    plt.plot(x, u, 'bo', xx, uu, 'k-', markersize=4, linewidth=1); plt.grid(True)
    plt.title(r"$u(x), N = " + str(N) + "$")
    
    uprime = np.exp(x) * (np.sin(5*x) + 5*np.cos(5*x))
    error = np.dot(D, u) - uprime
    plt.subplot(position=[0.55, 0.66 -0.4*(N==20), 0.31, 0.28])
    plt.plot(x, error, 'k-o', linewidth=1, markersize=4); plt.grid(True)
    plt.title(r"$\mathbf{D}_{N} \mathbf{u} - u^{\prime}(\mathbf{x}), N = " 
              + str(N) + "$");

